PostPublishPanel: return focus to element that opened the panel

oandregal opened the pull request:
Addresses the last bit of https://github.com/WordPress/gutenberg/issues/4187#issuecomment-427480786
~Depends on https://github.com/WordPress/gutenberg/pull/11584~
Closes #4187

## Description

This PR manages where focus goes after the PostPublishPanel is closed. At the moment, it's lost and just goes to the body element.

![peek 2018-11-09 13-33](https://user-images.githubusercontent.com/583546/48262800-22e21e80-e424-11e8-9365-349d707ae8a3.gif)

**Changes this PR introduces:**

- Uses the `withReturnFocus` HOC to return focus to the element that had it before opening the publish sidebar.
- Changes the Header settings not to be unmounted when the publish sidebar is opened. This requires us to hide the "Saving draft" button, because it is not completely hidden by the publish sidebar.
- Uses a single component as the Publish button, instead of the two we had. This helps in that the DOM node that opens the publish sidebar isn't changed by a new component. It does so by teaching the `PostPublishButton` how to behave like a button (publish directly) or a toggle (opens the publish sidebar instead).

## Questions

- [x] Should we deprecate `PostPublishToggle` in favor of the `PostPublishButton` component in this PR *We'll go with deprecating it and removing it in 5.1.*

oandregal committed:
Return focus to Publish... button

- Do not unmount Header settings components.
- Hide the "Save draft" button because it's not hidden by the
  PostPublishPanel slide-in sidebar.

nosolosw commented:
This approach gotchas:

* we need to hide the "Save draft" button as it isn't hidden completely by the PostPublishPanel.
* we need to refactor the publish toggle / button components to use only one element, otherwise, focus will be lost after changing from toggle to button (for ex: using the toggle to open the PostPublishPanel, which will get unmounted once the post is published in favor of the button).

oandregal committed:
Sort dependencies

oandregal committed:
Extract props

oandregal committed:
Toggle disabled if post is published

oandregal committed:
Toggle should be disabled if post is published

oandregal committed:
Add toggleProps to button

oandregal committed:
Inline text for toggle within the button component

oandregal committed:
Always render the button component

oandregal committed:
Use aria-disabled instead of disabled to avoid focus loss

oandregal committed:
Improve comments

oandregal committed:
Update snapshots

nosolosw commented:
So, against my original concerns, this PR has ended up being a fine approach. It is ready for review. The only thing I wonder is whether we should deprecate the toggle component. Normally, I'd do it, but given the point we're at in the release cycle I'd want to hear more opinions.

tofumatt reviewed (commented):
I'm unclear on whether #11584 needs to be merged before this will work as intended—is that the case?

Right now after closing the panel focus seems lost and is not returned to the element that opened it (the "Publish" button).

tofumatt commented:
For what it's worth: if we aren't using the `PostPublishToggle` component anywhere internally it seems the kind of thing we could deprecate as I'm unsure where else it would be being used…

But then we're past API freeze and should respect that, maybe we deprecate it as normal and remove it in a few versions, recognising that that'll be after WordPress 5.0.

oandregal committed:
Add deprecation notice in editor CHANGELOG

oandregal committed:
Add deprecation notice in plugin

oandregal committed:
Add deprecation in component

oandregal committed:
Add DotTip to PostPublishButton

nosolosw commented on docs/reference/deprecated.md:
```diff
@@ -1,5 +1,9 @@
 Gutenberg's deprecation policy is intended to support backwards-compatibility for releases, when possible. The current deprecations are listed below and are grouped by _the version at which they will be removed completely_. If your plugin depends on these behaviors, you must update to the recommended alternative before the noted version.
 
+## 5.1
```
I'm unsure how many plugin versions we're going to release before WordPress 5.0 is out. We are also in API-freeze mode, so I guess in this case makes sense to remove the component in the second planned version after public release.

tofumatt replied:
I think Gutenberg 4.3 and 4.4 will ship and 4.4 will be what merges into WordPress 5.0.

tofumatt replied:
So this should not be 5.1; those are WordPress version numbers. These deprecations are relevant to Gutenberg versions, where are separate.

nosolosw replied:
```suggestion
## 4.6
```

oandregal committed:
Fix deprecation warning in tests

nosolosw commented on packages/edit-post/src/components/header/index.js:
```diff
@@ -39,34 +39,32 @@ function Header( {
-						forceIsDirty={ hasActiveMetaboxes }
-						forceIsSaving={ isSaving }
+			<div className="edit-post-header__settings">
+				{ ! isPublishSidebarOpened && ( <PostSavedState
```
The `PostSavedState` component is not completely hidden by the slide-in `PostPublishPanel`, so we need to take care of this.

tofumatt replied:
Do we want to remove it from the DOM or just visually hide it? Should this comment be in the code?

tofumatt replied:
Also it'd be nice to me if the component was on a separate line from the paren; that was one of the tweaks I made before the force-push 😉 eg:

```suggestion
				{ ! isPublishSidebarOpened && (
					<PostSavedState
```

nosolosw commented on packages/edit-post/src/components/header/post-publish-button-or-toggle.js:
```diff
@@ -24,54 +24,54 @@ export function PostPublishButtonOrToggle( {
 
 	/**
-	 * We want to show a BUTTON when the post status is at the _final stage_
+	 * Conditions to show a BUTTON (publish directly) or a TOGGLE (open publish sidebar):
```
This big comment absorbs the previous two different existing comments.

nosolosw commented on packages/editor/src/components/post-publish-button/index.js:
```diff
@@ -66,17 +78,38 @@ export class PostPublishButton extends Component {
+		const toggleProps = {
+			'aria-disabled': isToggleDisabled,
+			'aria-expanded': isOpen,
+			className: 'editor-post-publish-panel__toggle',
```
Should we update this to be `editor-post-publish-toggle` instead? We can maintain both the existing and the new class until the toggle component is removed as to preserve compatibility. I could prepare a follow-up PR to this one with the changes if we wanted to do it.

nosolosw replied:
Or perhaps a better alternative would be to remove the `editor-post-publish-panel__toggle` class once the toggle component is removed.

tofumatt replied:
`editor-post-publish-panel__toggle` seems like the right name.

nosolosw commented on packages/editor/src/components/post-publish-button/index.js:
```diff
@@ -66,17 +78,38 @@ export class PostPublishButton extends Component {
 			>
-				<PublishButtonLabel forceIsSaving={ forceIsSaving } />
+				{ componentChildren }
+				<DotTip tipId="core/editor.publish">
```
The `DotTip` component was only visible for the toggle. This PR changes it and shows the tip no matter the button publish directly or opens the publish sidebar.

nosolosw commented:
@tofumatt It works for me: Firefox 63.0 and Chrome 70.0. in Linux (Ubuntu 16.04.5 LTS). As https://github.com/WordPress/gutenberg/pull/11584 is already merged, I've rebased this branch over master in case that's helpful (although it should not make any difference, though). Could you try again?

oandregal committed:
Update PostPublishButton e2e tests to use aria-disabled instead of disabled att

oandregal committed:
Update e2e test

This was relying on the preview button not being mounted
while the publish panel was opened. A better logic is to
check whether the actual publish panel is shown.

tofumatt committed:
Tweak comment

tofumatt commented on packages/edit-post/src/components/header/post-publish-button-or-toggle.js:
```diff
@@ -24,54 +24,54 @@ export function PostPublishButtonOrToggle( {
+	 * 	 Originally, we considered showing a button for pending posts that couldn't be published
+	 * 	 (for example, for an author with the contributor role). Some languages can have
+	 * 	 long translations for "Submit for review", so given the lack of UI real estate available
+	 * 	 we decided to take into account the viewport in that case.
```
I find this comment a bit confusing, but I know it was like that when you got here. I'm not really sure what:

> so given the lack of UI real estate available we decided to take into account the viewport in that case

means. 🤷‍♂️

If you can clarify it that'd be great, but no worries if not.

nosolosw replied:
Added a link to the PR that originated that change.

tofumatt commented on packages/editor/src/components/post-publish-button/index.js:
```diff
@@ -66,17 +78,38 @@ export class PostPublishButton extends Component {
-				<PublishButtonLabel forceIsSaving={ forceIsSaving } />
+				{ componentChildren }
+				<DotTip tipId="core/editor.publish">
+					{ __( 'Finished writing? That’s great, let’s get this published right now. Just click “Publish” and you’re good to go.' ) }
```
I find "Just" to be a weasel word and would like to remove it, but I know that's a bit out-of-scope for the PR. Just saying' 😆

tofumatt commented on packages/editor/src/components/post-publish-panel/test/toggle.js:
```diff
@@ -20,6 +20,9 @@ describe( 'PostPublishPanelToggle', () => {
 
 			expect( wrapper.prop( 'disabled' ) ).toBe( true );
+			expect( console ).toHaveWarnedWith(
+				'PostPublishPanelToggle is deprecated and will be removed from Gutenberg in 5.1. Please use PostPublishButton instead.'
```
This should be Gutenberg 4.6.

nosolosw replied:
```suggestion
				'PostPublishPanelToggle is deprecated and will be removed from Gutenberg in 4.6. Please use PostPublishButton instead.'
```

tofumatt commented on packages/editor/src/components/post-publish-panel/toggle.js:
```diff
@@ -25,6 +26,12 @@ export function PostPublishPanelToggle( {
 		( ! isPublishable && ! forceIsDirty );
 
+	deprecated( 'PostPublishPanelToggle', {
+		version: '5.1',
```
```suggestion
		version: '4.6',
```

nosolosw replied:
*Just* updated that myself :D

tofumatt reviewed (changes requested):
I'll try this again a bit later today after I'm back from squash, but for now I have a few change requests regarding the deprecations.

Any chance you could record a screencast of this working on your machine? It would help to see what you're seeing 😄

nosolosw committed:
Update version

nosolosw commented on packages/editor/src/components/post-publish-panel/toggle.js:
```diff
@@ -25,6 +26,12 @@ export function PostPublishPanelToggle( {
 		( ! isPublishable && ! forceIsDirty );
 
+	deprecated( 'PostPublishPanelToggle', {
+		version: '5.1',
```
```suggestion
		version: '4.6',
```

nosolosw committed:
Update packages/editor/src/components/post-publish-panel/toggle.js

tofumatt committed:
Place component in a separate line

Co-Authored-By: nosolosw <nosolosw@users.noreply.github.com>

nosolosw committed:
Update version number

oandregal committed:
Add comment

oandregal committed:
Update comment

nosolosw commented:
@tofumatt Added a GHIF in the description. After clicking the "Publish" button the first time I did everything by tabbing forward/backward and hitting the INTRO key. Hope that's helpful!

tofumatt commented on docs/reference/deprecated.md:
```diff
@@ -1,5 +1,9 @@
 Gutenberg's deprecation policy is intended to support backwards-compatibility for releases, when possible. The current deprecations are listed below and are grouped by _the version at which they will be removed completely_. If your plugin depends on these behaviors, you must update to the recommended alternative before the noted version.
 
+## 4.6
```
Best to keep this consistent with the rest of the file:

```suggestion
## 4.6.0
```

tofumatt commented on packages/edit-post/src/components/header/index.js:
```diff
@@ -39,34 +39,39 @@ function Header( {
-				<div className="edit-post-header__settings">
+			<div className="edit-post-header__settings">
+				{ ! isPublishSidebarOpened && (
+					// This button isn't completely hidden by the publish sidebar.
```
👍 

This is an awesome, super-helpful comment. Ace!

tofumatt committed:
Tweak comment

tofumatt reviewed (approved):
I tested this locally and now it's working perfectly. Thanks so much!

I left a few suggestions but once they're addressed 🚢

Thanks so much for this! Once it's merged, all the Gutenberg merge accessibility issues are covered! 🎉

tofumatt reviewed (changes requested):
Damn, looks like we shouldn't be deprecating stuff here:

https://github.com/WordPress/gutenberg/pull/11679#discussion_r232306860

Might be best to remove that for now. Really sorry, I thought deprecations would continue as normal, but seems like the mechanism will be different going forward, so for now maybe we shouldn't.

aduth commented:
>Damn, looks like we shouldn't be deprecating stuff here:
>
>#11679 (comment)

The changes of #11679 are not really relevant to speaking to whether or not a thing can be deprecated still. The changes there are merely a reflection of the fact that the dependency becomes unused once the current deprecations are removed.

tofumatt commented:
Right, but I thought @youknowriad was saying deprecations after 4.3 we don't know how we'll handle them.

Maybe I'm confused and these deprecations are fine? I'm confused 😅

youknowriad commented:
> Right, but I thought @youknowriad was saying deprecations after 4.3 we don't know how we'll handle them.

Since we said we're in API Freeze, yes I did say that and I think we should be more strict now to keep our engagements. We shouldn't deprecate anything impactful.

I think `PostPublishPanelToggle` is not going to impact any plugin though because it's something npm consumers will likely use but less WP plugins.

tofumatt commented:
Fair enough then; changed back to approved 😄

tofumatt commented:
We're actually going to merge this into 4.3; I'll take it over from here and tweak the deprecations. Thanks again for working on this 👍

tofumatt committed:
chore: Move deps to 4.5

aduth commented on packages/editor/src/components/post-publish-button/index.js:
```diff
@@ -66,17 +78,38 @@ export class PostPublishButton extends Component {
-				isPrimary
-				isLarge
-				onClick={ onClick }
-				disabled={ isButtonDisabled }
```
For what reason did we need to stop passing the `disabled` prop? While this and many other buttons in the header _appear_ as disabled, they don't act as disabled. You can immediately press "Publish" on a new post and the publish panel is (wrongly?) shown.

As it relates to focus return, it's not clear to me how it's relevant, since the panel would have presumably never been opened if it were disabled.

nosolosw replied:
Good catch. https://github.com/WordPress/gutenberg/pull/13194 should fix it.

The `disabled` property didn't play well with the `withFocusReturn` HOC, IIRC. I can dig up the exact details if necessary.
